### DISCUSSION TITLE
Scheduler endlessly prints "ValueError: source code string cannot contain null bytes"

### DISCUSSION THREAD

**Discussion by safronau** [apache/airflow#31519] [General] [answered]
### Apache Airflow version

2.6.1

### What happened

Scheduler endlessly prints:

```
app-airflow-scheduler-1  | Process ForkProcess-<XXX>:
app-airflow-scheduler-1  | Traceback (most recent call last):
app-airflow-scheduler-1  |   File "/usr/local/lib/python3.10/multiprocessing/process.py", line 314, in _bootstrap
app-airflow-scheduler-1  |     self.run()
app-airflow-scheduler-1  |   File "/usr/local/lib/python3.10/multiprocessing/process.py", line 108, in run
app-airflow-scheduler-1  |     self._target(*self._args, **self._kwargs)
app-airflow-scheduler-1  |   File "/home/airflow/.local/lib/python3.10/site-packages/airflow/dag_processing/manager.py", line 259, in _run_processor_manager
app-airflow-scheduler-1  |     processor_manager.start()
app-airflow-scheduler-1  |   File "/home/airflow/.local/lib/python3.10/site-packages/airflow/dag_processing/manager.py", line 493, in start
app-airflow-scheduler-1  |     return self._run_parsing_loop()
app-airflow-scheduler-1  |   File "/home/airflow/.local/lib/python3.10/site-packages/airflow/dag_processing/manager.py", line 639, in _run_parsing_loop
app-airflow-scheduler-1  |     self.start_new_processes()
app-airflow-scheduler-1  |   File "/home/airflow/.local/lib/python3.10/site-packages/airflow/dag_processing/manager.py", line 1092, in start_new_processes
app-airflow-scheduler-1  |     processor.start()
app-airflow-scheduler-1  |   File "/home/airflow/.local/lib/python3.10/site-packages/airflow/dag_processing/processor.py", line 213, in start
app-airflow-scheduler-1  |     self.import_modules(self.file_path)
app-airflow-scheduler-1  |   File "/home/airflow/.local/lib/python3.10/site-packages/airflow/dag_processing/processor.py", line 380, in import_modules
app-airflow-scheduler-1  |     _import_modules(file_path)
app-airflow-scheduler-1  |   File "/home/airflow/.local/lib/python3.10/site-packages/airflow/dag_processing/processor.py", line 366, in _import_modules
app-airflow-scheduler-1  |     for module in iter_airflow_imports(filepath):
app-airflow-scheduler-1  |   File "/home/airflow/.local/lib/python3.10/site-packages/airflow/utils/file.py", line 389, in iter_airflow_imports
app-airflow-scheduler-1  |     parsed = ast.parse(Path(file_path).read_bytes())
app-airflow-scheduler-1  |   File "/usr/local/lib/python3.10/ast.py", line 50, in parse
app-airflow-scheduler-1  |     return compile(source, filename, mode, flags,
app-airflow-scheduler-1  | ValueError: source code string cannot contain null bytes

```

It is deployed without zipped DAGS as descibed in #31039 
But `AIRFLOW__SCHEDULER__PARSING_PRE_IMPORT_MODULES=False` fixes it too.
But dags are exist.

### How to reproduce

I can not provide minimal code to reproduce.

### Operating System

linux/amd64

### Deployment

Docker-Compose

### Deployment details

source docker image: apache/airflow:2.6.1-python3.10

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template! If you are willing to raise PR to address this issue please do so, no need to wait for approval.

**Comment by potiuk** [comment, marked as answer]
Yes. Your DAG code contains null bytes. Probably you have a file that is not ignored with .airflowignore in DAG folder  - most likely a binary file (zip?) with .py extension. remove those. You can also try to remove all files, and add them one-by-one (or better bisect and add half files see if fail, when do, remove them and add half of those etc. This will allow you to pin-point the culprit quickly.

**Reply by safronau** [reply]
Thank you, it helps. In my case JAR file was there.
Also, I'd suggest to print the file name with null bytes. This will help to a lot of projects in future.
Thank you.

**Comment by jtommi** [comment]
@potiuk Thanks for the tip, but I still struggled to find the offending file.
So I added an except ValueError to find the file:
![image](https://github.com/apache/airflow/assets/33985391/75799cc2-6907-4cdf-b861-4f99dc7b8b9b)

Should I submit this as a pull request

**Reply by potiuk** [reply]
Sure! that will be a cool improvement. Tag me in it.
